keystore/src/entities/platform/generic/mls/enrollment.rs

```diff
@@ -65,42 +65,42 @@ impl EntityBase for E2eiEnrollment {
     }
 
     async fn save(&self, conn: &mut Self::ConnectionType) -> crate::CryptoKeystoreResult<()> {
-        use rusqlite::OptionalExtension as _;
         use rusqlite::ToSql as _;
 
         Self::ConnectionType::check_buffer_size(self.content.len())?;
 
         let transaction = conn.transaction()?;
-        let existing_rowid = transaction
-            .query_row("SELECT rowid FROM e2ei_enrollment WHERE id = ?", [&self.id], |r| {
-                r.get::<_, i64>(0)
-            })
-            .optional()?;
 
-        let row_id = if existing_rowid.is_some() {
-            return Err(CryptoKeystoreError::AlreadyExists);
-        } else {
-            let zb = rusqlite::blob::ZeroBlob(self.content.len() as i32);
-            let params: [rusqlite::types::ToSqlOutput; 2] = [self.id.to_sql()?, zb.to_sql()?];
-            transaction.execute("INSERT INTO e2ei_enrollment (id, content) VALUES (?, ?)", params)?;
-            transaction.last_insert_rowid()
-        };
-
-        let mut blob = transaction.blob_open(
-            rusqlite::DatabaseName::Main,
-            "e2ei_enrollment",
-            "content",
-            row_id,
-            false,
-        )?;
-
-        use std::io::Write as _;
-        blob.write_all(&self.content)?;
-        blob.close()?;
-
-        transaction.commit()?;
-
-        Ok(())
+        // Attempt to insert directly, handling conflicts as errors
+        let zb = rusqlite::blob::ZeroBlob(self.content.len() as i32);
+        let sql = "INSERT INTO e2ei_enrollment (id, content) VALUES (?, ?) RETURNING rowid";
+
+        let row_id_result: Result<i64, rusqlite::Error> =
+            transaction.query_row(sql, [self.id.to_sql()?, zb.to_sql()?], |r| r.get(0));
+
+        match row_id_result {
+            Ok(row_id) => {
+                // Open a blob to write the content data
+                let mut blob = transaction.blob_open(
+                    rusqlite::DatabaseName::Main,
+                    "e2ei_enrollment",
+                    "content",
+                    row_id,
+                    false,
+                )?;
+
+                use std::io::Write as _;
+                blob.write_all(&self.content)?;
+                blob.close()?;
+
+                transaction.commit()?;
+                Ok(())
+            }
+            Err(rusqlite::Error::SqliteFailure(_e, Some(ref msg))) if msg.contains("UNIQUE constraint failed") => {
```

Comment:
Can't we use direct error code checking instead of relying on the string representation?
I.e. see `SQLITE_CONSTRAINT_UNIQUE`: https://docs.rs/libsqlite3-sys/0.30.1/src/libsqlite3_sys/error.rs.html#260